text/0000-release-meta-spec-v2.md

```diff
@@ -0,0 +1,305 @@
+    the uploaded file (if it doesn't have the directory prefix, for example, or
+    is uploaded in some other archive format than zip).
+
+4.  PGXN Manager copies the `META.json` file, constructs the payload for
+    sighing, signs with its public key, then adds the `release` object to the
+    copied `META.json` file.
+
+5.   PGXN Manager publishes the distribution archive and the `META.json` file
+     to the root repository, along with an additional file that lists all the
+     releases of the distribution. These files would be:
+
+     *    Release list: `dist/pair.json`
+     *    0.1.7 release `META.json`: `dist/pair/0.1.7/META.json`
+     *    0.1.7 release zip file: `dist/pair/0.1.7/pair-0.1.7.zip`
+
+#### Installing a Release ####
+
+The steps for a client to find, download, and verify a PGXN release would be:
+
+1.  Using a valid PGXN mirror, assemble and fetch the release list for the
+    the extension, `dist/pair.json`.
+2.  Use the release list to determine the best version to install and
+    assemble its release `META.json` URI. The format is
+    `dist/{name}/{version}/META.json`; for the above example, that results
+    in `dist/pair/0.1.7/META.json`.
+3.  Fetch the release `META.json` file, read in the `releases/pgxn` object,
+    and use PGXN's public key to verify that it was signed by PGXN.
+4.  Decode the payload and use its `uri` field to download the release zip
+    file.
+5.  Compare the the `sha512` digest from the payload to a digest generated
+    from the downloaded zip file.
+6.  If they digests are the same, continue with building and installing.
+    Otherwise abort with an error.
+
+## Reference-level explanation
+
+This design ensures proper validation of PGXN releases by the following chain
+of trust:
+
+1.  A root key pair is maintained by PGXN, with the private key kept offline.
+2.  A release key pair is generated and signed by the private root key, with
+    the private key kept in an online vault accessible only to PGXN Manager.
+3.  The public keys for both keys are published by PGXN.
+4.  PGXN Manager uses the private release key to sign releases as described
+    above. The most important property in the signed payload is the list of
+    digests.
+5.  Clients can verify the signature with the public release and root keys.
+6.  With the data validated, the client can download and verify the release
+    file against a signed digest. In this manner, the authenticity of the
+    release file can be verified all the way to the root key.
+
+To support this infrastructure, PGXN Manager **MUST** be updated to properly
+generate and sign the payload and include it in the release `META.json` files.
+Clients **MUST** follow the [JWS validation steps].
+
+## Drawbacks
+
+*   This pattern could make it more difficult for clients to install code from
+    PGXN, especially if they incorrectly validate the signature.
+*   Some clients may choose not to implement the validation, potentially
```

Comment:
Should there be an option to install only trusted packages, or that's a client-specific consideration?

Reply:
Client-specific. My intention, once we have most stuff signed, would be for the official CLI to validate by default. But we have no day in how others want to use it.